app/javascript/controllers/autosave_controller.ts

```diff
@@ -0,0 +1,148 @@
+import invariant from 'tiny-invariant';
+import { httpRequest, ResponseError } from '@utils';
+import { z } from 'zod';
+
+import { ApplicationController } from './application_controller';
+import { AutoUpload } from '../shared/activestorage/auto-upload';
+
+const Gon = z.object({ autosave: z.object({ debounce_delay: z.number() }) });
+
+declare const window: Window & typeof globalThis & { gon: unknown };
+const { debounce_delay } = Gon.parse(window.gon).autosave;
+
+const AUTOSAVE_DEBOUNCE_DELAY = debounce_delay;
+const AUTOSAVE_TIMEOUT_DELAY = 60000;
+
+export class AutosaveController extends ApplicationController {
+  #abortController?: AbortController;
+  #latestPromise = Promise.resolve();
+  #needsRetry = false;
+
+  connect() {
+    this.#latestPromise = Promise.resolve();
+    this.onGlobal('autosave:retry', () => {
+      if (this.#needsRetry) {
+        this.enqueueAutosaveRequest();
+      }
+    });
+    this.on('change', (event) => this.onChange(event));
+  }
+
+  disconnect() {
+    this.#abortController?.abort();
+    this.#latestPromise = Promise.resolve();
+  }
+
+  onRetry(event: Event) {
+    const target = event.target as HTMLButtonElement;
+    const inputTargetSelector = target.dataset.inputTarget;
+    if (inputTargetSelector) {
+      const target =
+        this.element.querySelector<HTMLInputElement>(inputTargetSelector);
+      if (
+        target &&
+        target.type == 'file' &&
+        target.dataset.autoAttachUrl &&
+        target.files?.length
```

Comment:
Pour l'instant on ne retry pas autre chose que les attachments ?

Reply:
Il y a deux "retry” actuellement. Un local au champ que pour les attachments. Un global pour les autres types de champ. Je n'ai pas changé ça pour l'instant pour minimiser les changements. Mais je pense qu'on pourrait unifier les systèmes dans une autre itération.